### PR TITLE
Issue #3110: Bugfix in Checker.java removed English hardcoded msgs & Added new method to extract non-locale translations

### PR DESCRIPTION
**Pull request by MANISH-K-07**
Part of #3110 

The following methods required message keys : 

- processFiles (Exception msg)
- setupChild
- setCharset

Checker.java file now has only two more hardcoded strings left to completely be compatible with non-locale translations : 

1. destroy
2. processFiles (Error msg)

For maximum code optimization, I've designed a new method  `getLocalizedMessage` that takes in a String arg `messageKey` and returns another String `localizedMessage` that is extracted from _.properties_ files instead of returning _LocalizedMessage_ object, to make things even more simpler.

Thanks a lot @rdiachenko, for pointing this out coz, reused and optimized code is obviously the best !!

### CODE REVIEW

**Comment by MANISH-K-07** [comment]
The only english hardcoded strings I have left in Checker.java are : 

https://github.com/checkstyle/checkstyle/blob/646f49995cad29f62d489e14ba910788fc953aa6/src/main/java/com/puppycrawl/tools/checkstyle/Checker.java#L349-L360

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:69]: Why do we need it as a separate field? I'd suggest to use it in place.

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:107]: That's a lot of repetitive code.

1. Could you try to single out common part into a method to end up with something like this: `private final LocalizedMessage processFilesErrorMsg = localizedMessage("Checker.processFilesError");`.
2. Do we actually need those fields? Why can't we use inplace values? Considering, they are used once in rare cases when exception is thrown, I think we a good to have them closer to the usage.

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages_ru.properties
Comment (rdiachenko) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages_ru.properties:8]: Please update to: `не разрешен в качестве потомка в Checker`

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (MANISH-K-07) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:69]: The string literal "` `" has been used at a total of 4 places in the file. Apparently, we have a check called _MultipleStringLiterals_ that allows no repetitions (by default max allowed = 1).

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages_ru.properties
Comment (MANISH-K-07) [reply src/main/resources/com/puppycrawl/tools/checkstyle/messages_ru.properties]: Update done.
I have been using Google Translate to obtain the non-locale translations

**Comment by MANISH-K-07** [comment]
@rdiachenko ,
I am not that familiar with pitests side of our project. ( More into development and unit testing than mutation hunting 😅 )
There are few surviving mutations in pitest-commons 
https://github.com/checkstyle/checkstyle/actions/runs/8274073597/job/22638881050?pr=14652#step:6:476
Could you please guide me on how to deal with these ?

**Comment by Lmh-java** [comment]
@MANISH-K-07 I guess it is basically because you've introduced a new method (`getLocalizedMessage()`), but you don't have any tests for it. Therefore, some mutations for it in Pitest is not killed.

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties
Comment (Lmh-java) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties:7]: Hi, could you please update Checker.setupChildNotAllowed to `不允许作为Checker的子模块`. Since this would make more sense in Mandarin 😊

**Comment by MANISH-K-07** [comment]
> @MANISH-K-07 I guess it is basically because you've introduced a new method (`getLocalizedMessage()`), but you don't have any tests for it. Therefore, some mutations in Pitest are not killed.

I actually had the same surviving mutations during first push itself i.e. before introducing the newly designed method. So, I don't think the Pitest failure is coz of the `getLocalizedMessage` method. I suspect the fault here being with call to getMessage().
Thanks for your opinion anyways :)

Also, regarding the Chinese msg change, I'll get it done as soon as I get on my laptop

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:69]: 1. Could you make it `private` since it is used within this class only?
2. Please rename it as `WHITESPACE_SEPARATOR`
3. Update comment as follows: `Separator to use in string concatination.`

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:191]: Why do we need it as a separate variable? Can we use it inplace `throw new IllegalStateException(getLocalizedMessage("Checker.destroy"), ex);`?

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:664]: Please make it lowercase `Localized`

**Comment by rdiachenko** [comment]
> @rdiachenko , I am not that familiar with pitests side of our project. ( More into development and unit testing than mutation hunting 😅 ) There are few surviving mutations in pitest-commons https://github.com/checkstyle/checkstyle/actions/runs/8274073597/job/22638881050?pr=14652#step:6:476 Could you please guide me on how to deal with these ?

Some (not all) of your changes led to the following surviving mutations:
```
+  <mutation unstable="false">
+    <sourceFile>Checker.java</sourceFile>
+    <mutatedClass>com.puppycrawl.tools.checkstyle.Checker</mutatedClass>
+    <mutatedMethod>destroy</mutatedMethod>
+    <mutator>org.pitest.mutationtest.engine.gregor.mutators.NonVoidMethodCallMutator</mutator>
+    <description>removed call to com/puppycrawl/tools/checkstyle/Checker::getLocalizedMessage</description>
+    <lineContent>final String destroyMessage = getLocalizedMessage(&quot;Checker.destroy&quot;);</lineContent>
+  </mutation>
+
+  <mutation unstable="false">
+    <sourceFile>Checker.java</sourceFile>
+    <mutatedClass>com.puppycrawl.tools.checkstyle.Checker</mutatedClass>
+    <mutatedMethod>destroy</mutatedMethod>
+    <mutator>org.pitest.mutationtest.engine.gregor.mutators.experimental.ArgumentPropagationMutator</mutator>
+    <description>replaced call to com/puppycrawl/tools/checkstyle/Checker::getLocalizedMessage with argument</description>
+    <lineContent>final String destroyMessage = getLocalizedMessage(&quot;Checker.destroy&quot;);</lineContent>
+  </mutation>
+
+  <mutation unstable="false">
+    <sourceFile>Checker.java</sourceFile>
+    <mutatedClass>com.puppycrawl.tools.checkstyle.Checker</mutatedClass>
+    <mutatedMethod>processFiles</mutatedMethod>
+    <mutator>org.pitest.mutationtest.engine.gregor.mutators.NonVoidMethodCallMutator</mutator>
+    <description>removed call to com/puppycrawl/tools/checkstyle/Checker::getLocalizedMessage</description>
+    <lineContent>final String processFilesErrorMessage = getLocalizedMessage(</lineContent>
+  </mutation>
+
+  <mutation unstable="false">
+    <sourceFile>Checker.java</sourceFile>
+    <mutatedClass>com.puppycrawl.tools.checkstyle.Checker</mutatedClass>
+    <mutatedMethod>processFiles</mutatedMethod>
+    <mutator>org.pitest.mutationtest.engine.gregor.mutators.experimental.ArgumentPropagationMutator</mutator>
+    <description>replaced call to com/puppycrawl/tools/checkstyle/Checker::getLocalizedMessage with argument</description>
+    <lineContent>final String processFilesErrorMessage = getLocalizedMessage(</lineContent>
+  </mutation>
```
It means, that there are no real tests for code you introduced. To kill survivors you need to add such tests to cover new logic with getting localized message. To check whether pitest passes on localhost against specific pitest profile you may run this: `.ci/pitest.sh pitest-common`.

If you are not familiar with pitest at all, I'd suggest to start with this resource https://pitest.org

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (MANISH-K-07) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java]: Done..
Actually, I left the field `public static` so that it can be called later when required somewhere else while I work on rest of the files of this issue. I though this way, we could avoid multiple private field declarations for every class :)

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (MANISH-K-07) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java]: Done..
Even more optimization, that's fantastic

**Comment by MANISH-K-07** [comment]
@rdiachenko ,
I have browsed through https://pitest.org/ for the past few hours and now have a basic understanding of PIT.
Looks like we will need to cover tests that fail when call to getLocalizedMessage() is removed or replaced with arg for two places.

I will work on covering tests and send a follow-up PR as soon as this is merged. As this PR introduces a new method, I didn't wanna complicate things here :)

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages.properties
Comment (rdiachenko) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages.properties:1]: Please remove all commented lines and add them as part of a new PR when you work on it.

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java]: > later when required somewhere else

This "later" rarely comes in practice. Better and easier to apply restrictions now and loosen them later if needed than the other way around.

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages.properties
Comment (MANISH-K-07) [reply src/main/resources/com/puppycrawl/tools/checkstyle/messages.properties]: Done..
Archived on local and removed for remote branch.

**Review by rdiachenko — APPROVED** [review]
lgtm

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:72]:
```suggestion
    private static final String SINGLE_SPACE = " ";
```
We have no need to be ambiguous here. "Whitespace" could be multiple spaces, tabs, newlines, carriage returns, form feeds, vertical tabs, etc (or any combination thereof).

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Minor item:

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (MANISH-K-07) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:72]: I have originally used var `SINGLE_WHITESPACE` but later changed it after review https://github.com/checkstyle/checkstyle/pull/14652#discussion_r1524076122.

Done now.... re-changed it after re-review 😄

Comment (MANISH-K-07) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java]: I have originally used var `SINGLE_WHITESPACE` but later changed it after review https://github.com/checkstyle/checkstyle/pull/14652#discussion_r1524076122.

Done now.... re-changed it after re-review 😄

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages.properties
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages.properties:5]: All of them should have `{0}` to clearly define that some value will be there

**Review by romani — CHANGES_REQUESTED** [review]
Major item:

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java]: If we do https://github.com/checkstyle/checkstyle/pull/14652#discussion_r1526999404 this constant is not required at all

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java]: Yep, please do this @MANISH-K-07

Comment (MANISH-K-07) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java]: Sure.. already on it :)

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages.properties
Comment (MANISH-K-07) [reply src/main/resources/com/puppycrawl/tools/checkstyle/messages.properties]: Done and rebased....

Hacked the new method `getLocalizedMessage` to take another argument _**Varargs**_. Also, modified the driver codes to take params. There weren't much major changes required here bcoz `getMessage` already calls format on args if varargs is not null, just like our getCheckMessage :)

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:485]: All should be in message, nobody limit you to have `{1}`.
All should be in parameters of message. No any concatenation of content in code.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java
Comment (MANISH-K-07) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:485]: >  I was trying a rather different approach with one of the exceptions but there were conflicts

@romani , 
Actually, the change you suggested now is what I tried previously. I removed call to getMessage from hardcoded string and pointed it to .properties file as {1}
However, there were conflicts among two ci.
The `getLocalizedMessage` takes varargs as arguments, so we need to pass param Obj but the _ex.getMessage()_ returns a string (not a string object). So, without casting, we get a survivng error in `Checker / check (nullness-optional-interning) (pull_request)`
When we cast to type object (or even use string constructor invocation), we have violations with idea inspections saying the cast is redundant. I did not see a simpler way here to tackle both these ci other than pulling out ex.getMessage()
If there is any better way then I am all ears :)

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:485]: If static analysis tools blocks to make better code , we do suppression of them.

Comment (MANISH-K-07) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:485]: I kept seeing in many other cases that the maintainers generally recommend not to extend surviving errors. So, I thought we won't be allowing it here too.

> If static analysis tools blocks to make better code , we do suppression of them.

One suppression coming right up....

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:485]: Checker is not fully enforced tool, I would rather sacrifice it, in favor all other, we will deal with it later on.

Main point that whole message must be in property file. All other are internal details that we can change at any time. Message is property file is kind of a contract to all users, it should not change without notifying them, user should be able to translate and change it if required.

Comment (MANISH-K-07) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java]: Please ignore this push. It is to obtain patch

**Review by romani — APPROVED** [review]
ok to merge

**Comment by romani** [comment]
@rdiachenko , @nrmancuso , @rnveach , FYI, better to have Checker suppression extension than not good code.

**Pull request merged**
